pkg/controller/controller_utils.go

```diff
@@ -402,3 +405,129 @@ func ClusterAPIProviderStatusFromClusterStatus(clusterStatus *clusteroperator.Cl
+			APIVersion: clusteroperator.SchemeGroupVersion.String(),
+			Kind:       "MachineSetProviderConfigSpec",
+		},
+		MachineSetSpec: *machineSetSpec,
+	}
+	serializer := jsonserializer.NewSerializer(jsonserializer.DefaultMetaFactory, api.Scheme, api.Scheme, false)
+	var buffer bytes.Buffer
+	err := serializer.Encode(msProviderConfigSpec, &buffer)
+	if err != nil {
+		return nil, err
+	}
+	return &runtime.RawExtension{
+		Raw: buffer.Bytes(),
+	}, nil
+}
+
+// PopulateMachineSpec ensures that the MachineSetSpec we use for the machine spec provider config is fully populated with defaults and calculated values based on Cluster Operator specific logic. This can be used both on individual machines, as well as on the MachineTemplateSpec used in MachineSets and MachineDeployments.
+func PopulateMachineSpec(machineSpec *clusterapi.MachineSpec, clusterSpec *clusteroperator.ClusterSpec, clusterVersion *clusteroperator.ClusterVersion, mLog log.FieldLogger) error {
+	var msSpec *clusteroperator.MachineSetSpec
+	var err error
+	if machineSpec.ProviderConfig.Value == nil {
+		msSpec = &clusteroperator.MachineSetSpec{}
+	} else {
+		msSpec, err = MachineSetSpecFromClusterAPIMachineSpec(machineSpec)
+		if err != nil {
+			return err
+		}
+	}
+
+	if msSpec.VMImage.AWSImage == nil {
+		mLog.Debugf("no VMImage set, using cluster version")
+
+		vmImage, err := getImage(clusterSpec, clusterVersion)
+		if err != nil {
+			return err
+		}
+		msSpec.VMImage = *vmImage
+		mLog.Debugf("machine spec VMImage set to: %s", vmImage)
+	}
+
+	// use cluster defaults for hardware spec if unset:
+	hwSpec, err := ApplyDefaultMachineSetHardwareSpec(msSpec.Hardware, clusterSpec.DefaultHardwareSpec)
+	if err != nil {
+		return err
+	}
+	msSpec.Hardware = hwSpec
+	mLog.Debug("machine spec hardware set to: %v", msSpec.Hardware.AWS)
+
+	// Copy cluster hardware onto the provider config as well. Needed when deleting a cluster in the actuator.
+	if (msSpec.ClusterHardware == clusteroperator.ClusterHardwareSpec{}) {
+		mLog.Debugf("cluster hardware not set, copying: %v", clusterSpec.Hardware)
+		msSpec.ClusterHardware = clusterSpec.Hardware
+	}
+
+	providerConfig, err := ClusterAPIMachineProviderConfigFromMachineSetSpec(msSpec)
+	if err != nil {
+		return err
+	}
+	machineSpec.ProviderConfig.Value = providerConfig
+	return nil
+}
+
+// getImage returns a specific image for the given machine and cluster version.
+func getImage(clusterSpec *clusteroperator.ClusterSpec, clusterVersion *clusteroperator.ClusterVersion) (*clusteroperator.VMImage, error) {
+	if clusterSpec.Hardware.AWS == nil {
+		return nil, fmt.Errorf("no AWS hardware defined for cluster")
+	}
+
+	if clusterVersion.Spec.VMImages.AWSImages == nil {
+		return nil, fmt.Errorf("no AWS images defined for cluster version")
+	}
+
+	for _, regionAMI := range clusterVersion.Spec.VMImages.AWSImages.RegionAMIs {
+		if regionAMI.Region == clusterSpec.Hardware.AWS.Region {
+			return &clusteroperator.VMImage{
+				AWSImage: &regionAMI.AMI,
```

Comment:
Copy `regionAMI.AMI` to a new string and use a pointer to that. It would be really hard to track down if somewhere the value stored in this pointer where modified and it mutated a ClusterVersion stored in the lister.